gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/DiagnosticsManifestPlugin.java

```diff
@@ -0,0 +1,190 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.dist.service;
+
+import com.palantir.gradle.dist.tasks.CreateManifestTask;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ArtifactView;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.transform.CacheableTransform;
+import org.gradle.api.artifacts.transform.InputArtifact;
+import org.gradle.api.artifacts.transform.TransformAction;
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.artifacts.transform.TransformParameters;
+import org.gradle.api.attributes.LibraryElements;
+import org.gradle.api.attributes.Usage;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.CacheableTask;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DiagnosticsManifestPlugin implements Plugin<Project> {
+
+    public static final String mergeDiagnosticsJson = "mergeDiagnosticsJson";
+
+    /**
+     * This plugin uses Gradle's Artifact Transforms to extract a single file from all the jars on the classpath
+     * (https://docs.gradle.org/current/userguide/artifact_transforms.html).
+     * All we do is tell gradle what function can turn 'usage=jar' into 'usage=(our thing)', and then declare a
+     * a view using 'usage=(our thing)' and it'll just run the {@link ExtractSingleFile} transform
+     * to process the jars and extract the stuff we want! Crucially, this is all cached _beautifully_.
+     */
+    @Override
+    public void apply(Project project) {
+        String fileToExtract = "sls-manifest/diagnostics.json";
+        String attribute = "extracted-" + fileToExtract;
+
+        project.getDependencies().registerTransform(ExtractSingleFile.class, details -> {
+            details.getParameters().getPathToExtract().set(fileToExtract);
+
+            // this USAGE_ATTRIBUTE is already present on everything, so gradle can figure out how to transform to our
+            // attribute value
```

Comment:
Usage is defined here https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java